src/main.rs

```diff
@@ -1047,3 +1113,359 @@ fn main() -> Result<(), Box<dyn std::error::Error + 'static>> {
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn byteset_is_disjoint() {
+        let not_00 = &ByteSet::not(0x00);
+        let not_255 = &ByteSet::not(0xFF);
+        let is_00 = &ByteSet::is(0x00);
+        let is_255 = &ByteSet::is(0xFF);
+        assert!(!ByteSet::is_disjoint(not_00, not_00));
+        assert!(!ByteSet::is_disjoint(not_00, not_255));
+        assert!(ByteSet::is_disjoint(not_00, is_00));
+        assert!(!ByteSet::is_disjoint(not_00, is_255));
+        assert!(ByteSet::is_disjoint(is_00, is_255));
+    }
+
+    fn accepts(d: &Decoder, input: &[u8], tail: &[u8], expect: Value) {
+        let mut stack = Vec::new();
+        let (val, remain) = d.parse(&mut stack, input).unwrap();
+        assert_eq!(val, expect);
+        assert_eq!(remain, tail);
+    }
+
+    fn rejects(d: &Decoder, input: &[u8]) {
+        let mut stack = Vec::new();
+        assert!(d.parse(&mut stack, input).is_none());
+    }
+
+    #[test]
+    fn compile_fail() {
+        let f = Format::Fail;
+        let d = Decoder::compile(&f, None).unwrap();
+        rejects(&d, &[]);
+        rejects(&d, &[0x00]);
+    }
+
+    #[test]
+    fn compile_empty() {
+        let f = Format::Empty;
+        let d = Decoder::compile(&f, None).unwrap();
+        accepts(&d, &[], &[], Value::Unit);
+        accepts(&d, &[0x00], &[0x00], Value::Unit);
```

Comment:
I usually prefer to separate out independent test cases into separate test functions so they can fail independently, but it’s not a huge issue (happy to fix that myself). Good to have some tests for this.